### PR TITLE
DAOS-5912 rebuild: restore pool map if extend/drain fail (#11411)

### PR DESCRIPTION
**Pull request by wangdi1**
restore the pool map if extend/drain fail.
Add tests to verify it.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by jolivier23 — CHANGES_REQUESTED** [review]
raft submodule update is intentional?

**Review by liw — APPROVED** [review]

File: src/pool/srv_pool_map.c [hunk 1]
```diff
@@ -177,20 +177,30 @@ update_one_tgt(struct pool_map *map, struct pool_target *target,
 			D_INFO("Skip EXCLUDE_OUT DOWNOUT "DF_TARGET"\n",
 			       DP_TARGET(target));
 			break;
-		case PO_COMP_ST_UP:
 		case PO_COMP_ST_UPIN:
-		case PO_COMP_ST_NEW:
 			D_ERROR("Can't EXCLUDE_OUT non-down "DF_TARGET"\n",
 				DP_TARGET(target));
 			return -DER_INVAL;
+		case PO_COMP_ST_NEW:
```
Comment (liuxuezhao) [inline src/pool/srv_pool_map.c:184]: here the code is different with master barnch that cannot POOL_EXCLUDE_OUT NEW comp, is it intended to make this difference for 2.2?

**Review by liuxuezhao — APPROVED** [review]

**Comment by wangdi1** [comment]
please review. thanks

File: src/pool/srv_pool_map.c (hunk 1 above)
Comment (wangdi1) [reply src/pool/srv_pool_map.c:184]: yes, 2.4 change the NEW/UP target status for extending.

**Review by liuxuezhao — APPROVED** [review]

**Review by liuxuezhao — DISMISSED** [review]

**Comment by wangdi1** [comment]
all tests passed. ping reviewers.

**Comment by kprantis** [comment]
Reviewers, please prioritize this review.

**Comment by wangdi1** [comment]
ping

**Review by liuxuezhao — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

File: src/include/daos_srv/rebuild.h [hunk 2]
```diff
@@ -21,15 +21,15 @@
  * associated targets
  */
 typedef enum {
-	RB_OP_FAIL,
+	RB_OP_EXCLUDE,
```
Comment (ashleypittman) [inline src/include/daos_srv/rebuild.h:24]: @jolivier23 not for this PR obviously but would this be better as an X macro in master?

Comment (wangdi1) [reply src/include/daos_srv/rebuild.h:24]: this is not RPC code, but used by pool and rebuild module to tell different rebuild options.

**Pull request merged**